apps/initializer/lib/initializer.ex

```diff
@@ -30,29 +31,47 @@ defmodule Initializer do
           Map.new(init_arg)
           |> Map.put(:supervisor_ref, supervisor_ref)
 
-        case on_start(state) do
-          {:ok, new_state} -> {:ok, new_state}
+        {:ok, state, {:continue, :init}}
+      end
+
+      def handle_continue(:init, state) do
+        case do_on_start(state) do
+          {:ok, new_state} -> {:noreply, new_state}
           {:error, reason} -> {:stop, reason}
         end
       end
 
       def handle_info({:DOWN, supervisor_ref, _, _, _}, %{supervisor_ref: supervisor_ref} = state) do
-        retry with: constant_backoff(100) |> Stream.take(10), atoms: [false] do
-          Process.whereis(unquote(supervisor)) != nil
-        after
-          _ ->
+        case do_wait_for_supervisor(supervisor_ref) do
+          {:ok, _} ->
             supervisor_ref = setup_monitor()
             state = Map.put(state, :supervisor_ref, supervisor_ref)
 
             case on_start(state) do
               {:ok, new_state} -> {:noreply, state}
               {:error, reason} -> {:stop, reason, state}
             end
-        else
-          _ -> {:stop, "Supervisor not available", state}
+
+          _ ->
+            {:stop, "Supervisor not available", state}
         end
       end
 
+      @retry with: constant_backoff(100) |> take(10)
+      defp do_wait_for_supervisor(supervisor) do
```

Comment:
<nitpick> i don't think `do_wait_for_supervisor/1` has a corresponding `wait_for_supervisor/1` so maybe rename this to avoid confusion about additional callbacks that don't need to be implemented? </nitpick>